### DISCUSSION TITLE
Testing the digest inside a KDF.

### DISCUSSION THREAD

**Discussion by slontis** [openssl/openssl#24904] [General Discussion]
Some KDF objects have a MAC inside them. The Digest is set via a ossl_param inside ossl_prov_macctx_load_from_params().
There doesnt seem to be a way to extract the EVP_MD* back out of the EVP_MAC_CTX.
Which means you would have to do a EVP_MD_fetch() inside the kdf 
e.g.
````
    if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_DIGEST)) != NULL) {
        md = EVP_MD_fetch(NULL, p->data, NULL) ;
        if (EVP_MD_is_a(md, "SHA-256"))
            do something..
         EVP_MD_free(md);
         if (!ossl_prov_macctx_load_from_params(&ctx->P_hash, params,
                                                   OSSL_MAC_NAME_HMAC, .....)
````
Is there a better way to do this?

EVP_MAC_md_is_a() ?

**Comment by slontis** [comment]
The MAC cipher has a similar problem.
This probably also applies to EVP_KDF

**Reply by t8m** [reply]
Hmm, we could add a generic OSSL_ALGORITHM_NAME_eq() function that would just query the name maps.

**Comment by slontis** [comment]
i was looking at evp_is_a() and wondering if that was nice or not.. prov would be null in that case?

**Reply by t8m** [reply]
IMO that needs a fetched algorithm. But assuming the digest was already fetched by the internal fetch inside the KDF or MAC the name map will already contain the full name mapping, so it would be enough to just directly query the name map.

Now the problem is that the function would not work properly if the algorithm was not ever fetched before - not sure such semantics would be acceptable for a public API function.

**Reply by t8m** [reply]
We could have a tri-state return -1 - neither of the names is known to the libctx yet 0 - the names are not equivalent or one of the names is not known yet 1 - the names are equivalent.

**Comment by paulidale** [comment]
The KDF has a reference to an opaque MAC which in turn has a reference to an opaque digest.

I cannot think of a better approach than parsing the OSSL_PARAMs given to the KDF.

**Reply by slontis** [reply]
What does parsing mean, you mean doing a EVP_MD_fetch() ?

**Reply by paulidale** [reply]
Worst case, yes.

It might be possible to get away with just doing a `strcmp(3)` or `strcasecmp(3)` or  to check the name -- this is dangerous because we've got aliases which might mean `strstr(3)` or `strcasestr(3)` instead.  Even then a third party provider could invent a new name (which would then be an alias for all providers).

I'd probably play it safe and you the code snippet you first presented.  Maybe include the properties for the query too (not sure if this is going to make a difference or not).

**Comment by slontis** [comment]
So in the PR I am looking at it uses ossl_prov_digest_load_from_params() to do this.. and stores the PROV_DIGEST into the ctx..

To do it without storing in the ctx requires
````
if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_DIGEST)) != NULL) {
    PROV_DIGEST digest;
    const EVP *md;

    memset(&digest, 0, sizeof(digest));
    ossl_prov_digest_load_from_params(&digest, params, ctx); //needs error checking
    md = ossl_prov_digest_md()
     if (EVP_MD_is_a(md, "SHA-256"))
            do something.
    ossl_prov_digest_reset(&digest);
    
     if (!ossl_prov_macctx_load_from_params(&ctx->P_hash, params,
}
````

**Reply by slontis** [reply]
https://github.com/openssl/openssl/blob/6524abe289cf76ac9dd2e9fe6d7ca55027c72812/providers/implementations/kdfs/tls1_prf.c#L113

**Reply by paulidale** [reply]
It's not 100% but it's good enough (using PROV_DIGEST i.e.).

The not 100% is if a user passes a digest param but no MAC param, you'll false positive.  I don't think it's worth worrying about this case.
